File: src/helpers/RecordsProcessor.js
const avro = require('avsc')

const logger = require('./Logger.js')

// Map records to decode Avro and return data in desired output format to firehose.
const processRecords = function (schema, records, outputFormat) {
  let success = 0
  let failure = 0
  const type = avro.Type.forSchema(schema)
  const output = records.map((record) => {
    const decodedData = decodeAvro(type, record.data)
    if (!decodedData) {
      failure++
      return {
        recordId: record.recordId,
        result: 'ProcessingFailed',
        data: record.data
      }
    }

    let resultString = ''
    if (outputFormat === 'csv') {
      Object.values(decodedData).forEach((value) => {
        if (value === null) {
          value = ''
        } else if (typeof value === 'string') {
          value = value.replace('|', '\\|')
        }
        resultString += (value + '|')
      })
      resultString = resultString.slice(0, -1)
    } else {
      resultString = JSON.stringify(decodedData)
    }

    success++
    return {
      recordId: record.recordId,
      result: 'Ok',
      data: Buffer.from(resultString).toString('base64')
    }
  })
  logger.info(`Processing completed.  Successful transformations -  ${success}.  Failed transformations - ${failure}.`)
  return output
}

const decodeAvro = function (type, record) {
  const decodedRecord = Buffer.from(record, 'base64')
  try {
    return type.fromBuffer(decodedRecord)
  } catch (e) {
    logger.error(`Decoding fatal error occurred: ${e.message}`, { debugInfo: e })
    return false
  }
}

module.exports = {
  processRecords,
  decodeAvro
}

File: src/helpers/records_processor.py
import os
import avro
import json

import avro.schema
from nypl_py_utils.functions.log_helper import create_log
from nypl_py_utils.classes.avro_client import AvroDecoder


class RecordsProcessor:
    def __init__(self):
        self.logger = create_log("records_processor")
        self.avro_decoder = AvroDecoder(os.environ["NYPL_DATA_API_BASE_URL"])

    def _format_result_string(self, output_format, decoded_record):
        if output_format == "csv" and isinstance(decoded_record, dict):
            result_string, formatted_value = "", ""
            for value in decoded_record.values():
                if isinstance(value, str):
                    formatted_value = value.replace("|", "\\|")
                result_string += f"{formatted_value}|"
            return result_string[:-1] + "\n"
        else:
            return json.dumps(decoded_record["data"])

    def process_records(self, schema, records, output_format):
        """
        Map records to decoded Avro and return data in desired output
        format to Firehose.
        """
        successes, failures = 0, 0
        output = []
        type = avro.schema.parse(schema).type

        for record in records:
            decoded_record = self.avro_decoder.decode_record(record)

            if decoded_record is None:
                # Unable to decode Avro record
                failures += 1
                output.append(
                    json.dumps(
                        {
                            "recordId": record["recordId"],
                            "result": "ProcessingFailed",
                            "data": record["data"],
                        }
                    )
                )
            else:
                result_string = self._format_result_string(
                    output_format, decoded_record
                )
                successes += 1
                output.append(
                    json.dumps(
                        {
                            "recordId": record["recordId"],
                            "result": "Ok",
                            "data": result_string.encode("base64"),
                        }
                    )
                )
        self.logger.info(
            f"Processing completed.  Successful transformations -  ${successes}.  Failed transformations - ${failures}."
        )
        return output
